src/k2/module/k23si/TxnManager.cpp

```diff
@@ -670,69 +678,110 @@ seastar::future<Status> TxnManager::_finalizedPIP(TxnRecord& rec) {
                     }
 
                     K2LOG_D(log::skvsvr, "Erasing txn record: {}", rec);
-                    _transactions.erase(rec.txnId);
+                    _transactions.erase(rec.mtr.timestamp);
                     return seastar::make_ready_future();
                 });
         });
     return seastar::make_ready_future<Status>(dto::K23SIStatus::OK);
 }
 
+// This helper generates the finalization requests for the given txn record
+auto _genFinalizeRequests(TxnRecord& rec) {
+    std::deque<std::tuple<dto::K23SITxnFinalizeRequest, dto::KeyRangeVersion>> requests;
+    const auto endAction = rec.state == dto::TxnRecordState::Committed ? dto::EndAction::Commit : dto::EndAction::Abort;
+    for (auto& [coll, rangeSet]: rec.writeRanges) {
+        for (auto& krv: rangeSet) {
+            requests.emplace_back(
+            dto::K23SITxnFinalizeRequest{
+                .pvid=krv.pvid,
+                .collectionName=coll,
+                .txnTimestamp=rec.mtr.timestamp,
+                .action=endAction
+            },
+            std::move(krv));
```

Comment:
Is it safe to move krv here while you are iterating over the set?

Reply:
Technically, yes. The KRV here comes from the tracked ranges in the TXNRecord. In this method, I convert these into the requests I want to send. I don't look at these any further, but it makes following the code difficult so I'll just copy it for now